src/trace/layers/mod.rs

```diff
@@ -10,30 +10,33 @@ pub mod hashed;
 pub mod weighted;
 pub mod unordered;
 
+use std::rc::Rc;
+use owning_ref::OwningRef;
+
 /// A collection of tuples, and types for building and enumerating them.
 ///
 /// There are some implicit assumptions about the elements in trie-structured data, mostly that
 /// the items have some `(key, val)` structure. Perhaps we will nail these down better in the
 /// future and get a better name for the trait.
-pub trait Trie  : ::std::marker::Sized {
+pub trait Trie<B>  : ::std::marker::Sized {
 	/// The type of item from which the type is constructed.
 	type Item;
 	/// The type of cursor used to navigate the type.
 	type Cursor: Cursor;
 	/// The type used to merge instances of the type together.
-	type MergeBuilder: MergeBuilder<Trie=Self>;
+	type MergeBuilder: MergeBuilder<B, Trie=Self>;
 	/// The type used to assemble instances of the type from its `Item`s.
-	type TupleBuilder: TupleBuilder<Trie=Self, Item=Self::Item>;
+	type TupleBuilder: TupleBuilder<B, Trie=Self, Item=Self::Item>;
 
 	/// The number of distinct keys, as distinct from the total number of tuples.
 	fn keys(&self) -> usize;
 	/// The total number of tuples in the collection.
 	fn tuples(&self) -> usize;
 	/// Returns a cursor capable of navigating the collection.
-	fn cursor(&self) -> Self::Cursor { self.cursor_from(0, self.keys()) }
+	fn cursor(&self, owned_self: OwningRef<Rc<B>, Self>) -> Self::Cursor { self.cursor_from(owned_self, 0, self.keys()) }
```

Comment:
Is there something that prevents this from being a `OwningRef<Rc<Self>, Self>`? I'm trying to track down the place where the need to have the parameter `B` comes in, and this is where I got. :)